libimage/push.go

```diff
@@ -29,28 +32,79 @@ func (r *Runtime) Push(ctx context.Context, source, destination string, options
 		options = &PushOptions{}
 	}
 
-	// Look up the local image.  Note that we need to ignore the platform
-	// and push what the user specified (containers/podman/issues/10344).
-	image, resolvedSource, err := r.LookupImage(source, nil)
+	// Push the single image
+	if !options.AllTags {
+
+		// Look up the local image.  Note that we need to ignore the platform
+		// and push what the user specified (containers/podman/issues/10344).
+		image, resolvedSource, err := r.LookupImage(source, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure we have a proper destination, and parse it into an image
+		// reference for copying.
+		if destination == "" {
+			// Doing an ID check here is tempting but false positives (due
+			// to a short partial IDs) are more painful than false
+			// negatives.
+			destination = resolvedSource
+		}
+
+		return pushImage(ctx, image, destination, options, resolvedSource, r)
+	}
+
+	// Below handles the AllTags option, for which we have to build a list of
+	// all the local images that match the provided repository and then push them.
+	//
+	// Start by making sure a destination was not specified, since we'll get that from the source
+	if len(destination) != 0 {
+		return nil, fmt.Errorf("`destination` should not be specified if using AllTags")
+	}
+
+	// Make sure the source repository does not have a tag
+	srcNamed, err := reference.ParseNormalizedNamed(source)
 	if err != nil {
 		return nil, err
 	}
+	if _, hasTag := srcNamed.(reference.NamedTagged); hasTag {
+		return nil, fmt.Errorf("can't push with AllTags if source tag is specified")
+	}
 
+	// Now list every image of that source repository
+	listOptions := &ListImagesOptions{}
+	listOptions.Filters = []string{fmt.Sprintf("reference=%s:*", source)}
+	logrus.Debugf("Finding all images for source %s", source)
+	srcImages, _ := r.ListImages(ctx, nil, listOptions)
+
+	// Push each tag for every image in the list
+	var byteManifest []byte
+	for _, img := range srcImages {
+		namedTagged, err := img.NamedTaggedRepoTags()
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range namedTagged {
+			destWithTag := fmt.Sprintf("%s:%s", source, n.Tag())
```

Comment:
`reference.WithTag(srcNamed, n.Tag()`.

But I suspect the whole tag listing thing might need reworking and then this becomes unnecessary.

(Hypothetically: List all images, obtain their `.Names()`, parse them using `ParseNormalizedNamed`, use those where `.Name()` matches the user’s input. It’s quite possible that libimage already implements something like that, @vrothberg would know.

Alternatively, it’s of course possible to use `ListImages` and then do an extra filtering step, but that seems conceptually redundant.)

Reply:
I looked at the WithTags option, but that returns a different type, so didn't use it. 

I looked through most of the libimage functions and nothing caught my eye as a better option. I've been having a difficult time juggling the different ways to refer to a single image and its tag.